Apache Ignite Dataset

dmitrievanthony opened the pull request:
This is a proposal to add `IgniteDataset` that allows to work with [Apache Ignite](https://ignite.apache.org/). 

Apache Ignite is a memory-centric distributed database, caching, and processing platform for transactional, analytical, and streaming workloads, delivering in-memory speeds at petabyte scale. This proposal is a part of a more global initiative to implement so called "TensorFlow on Apache Ignite" ([IGNITE-8335](https://issues.apache.org/jira/browse/IGNITE-8335), [Design Document](https://docs.google.com/document/d/1jROIahK1rc7bSgOvhJhfpMqIGvht_IE8zn5NAt6x8ks)).

The integration is based on Apache Ignite [Binary Client Protocol](https://apacheignite.readme.io/v2.6/docs/binary-client-protocol) and TensorFlow [tf.data.Dataset](https://www.tensorflow.org/guide/datasets). More information about supported features you can find in [README.md](https://github.com/dmitrievanthony/tensorflow/blob/apache-ignite-dataset/tensorflow/contrib/ignite/README.md) of this module.

Tests have also been added. They use docker to hide configuration complexity, so that the implemented functionality can be tested quite simply by manual run.

**It's a copy of #21853, because previous request has accidentally been closed because I cleaned up merge commits from the branch.**

dmitrievanthony committed:
Add IgniteDataset that allows to work with Apache Ignite.

dmitrievanthony committed:
Remove duplicated header from README.md.

dmitrievanthony committed:
Update after review: change 'ignite' namespace to 'tensorflow', rename variables to satisty code style, use pointers instead of references.

dmitrievanthony committed:
Update README.md.

dmitrievanthony committed:
Fix pylint checks, fix VS compilation issue.

dmitrievanthony committed:
Fix code style.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    uint8_t* initial_ptr = ptr_;
+    std::vector<int32_t> types;
```
By the above comment, I think this vector is unused. However, since the data could have types that don't match the schema, we need some validation here, to check that `tensors` match the output types and shapes implied by the schema.

dmitrievanthony replied:
Thanks, as I wrote in comment above I forgot about it. I've added `types` filling and correspondent check. Please, see `CheckTypes` method.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+
+    remainder_ -= (ptr_ - initial_ptr);
+
+    out_tensors->resize(tensors.size());
```
Performance: using `out_tensors->push_back()` is more efficient than resizing and default-constructing the `Tensor` objects in the vector. This might require inverting the permutation.

dmitrievanthony replied:
Done. Inversion of permutation I placed on earlier stage (into `ignite_dataset_ops.cc`).

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+    remainder_ -= (ptr_ - initial_ptr);
+
+    out_tensors->resize(tensors.size());
+    for (int32_t i = 0; i < tensors.size(); i++)
```
Use `size_t` when ranging over a vector's size.

dmitrievanthony replied:
Thanks, done.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+    if (header == string_val) {
+      int32_t length;
+      TF_RETURN_IF_ERROR(client_->ReadInt(&length));
+      uint8_t* err_msg_c = new uint8_t[length];
```
This will leak if the following `client_->ReadData()` call fails. Instead of using `new[]`, do something like this:

```c++
string err_msg;
err_msg.resize(length);
TF_RETURN_IF_ERROR(client_->ReadData(err_msg.data(), length));
```

Please fix this pattern throughout the code.

dmitrievanthony replied:
Looks like I can't use `err_msg.data()` because it returns `const` object, but I need to fill it in `ReadData`. I added `gtl::MakeCleanup`, so I think leak is avoided.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+}
+
+Status IgniteDatasetIterator::ScanQuery() {
+  TF_RETURN_IF_ERROR(client_->WriteInt(25));                   // Message length
```
Move the magic number 25 to a named constant, and use it consistently in here.

dmitrievanthony replied:
I replaced all magic numbers by constants (where it's appropriate).

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+  int64_t stop = std::chrono::duration_cast<std::chrono::milliseconds>(
+                     std::chrono::system_clock::now().time_since_epoch())
+                     .count();
+  ;
```
Delete this empty statement.

dmitrievanthony replied:
Deleted.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+  page_ = std::unique_ptr<uint8_t>(new uint8_t[remainder_]);
+  ptr_ = page_.get();
+
+  int64_t start = std::chrono::duration_cast<std::chrono::milliseconds>(
```
Optional: Use `tensorflow::Env::NowSeconds()` here and below.

dmitrievanthony replied:
Thanks for advice. Replaced with `Env::Default()->NowMicros()`.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+                      std::chrono::system_clock::now().time_since_epoch())
+                      .count();
+
+  TF_RETURN_IF_ERROR(client_->ReadData(ptr_, remainder_));
```
If this fails, will `remainder_`, `page_` and `ptr_` be valid on a subsequent call to `GetNextInternal()`?

dmitrievanthony replied:
Yes, you are right. I've added field `valid_state_` that is set to `false` if something goes wrong. After that Iterator stops trying to get next object and will return error immediately.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+  int32_t row_cnt;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&row_cnt));
+
+  remainder_ = res_len - 17;
```
Where does 17 come from? Use named constants for quantities like this to make the code easier to follow.

dmitrievanthony replied:
Replaced by correspondent constant.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+  TF_RETURN_IF_ERROR(client_->ReadInt(&row_cnt));
+
+  remainder_ = res_len - 17;
+  page_ = std::unique_ptr<uint8_t>(new uint8_t[remainder_]);
```
This code (to the bottom of the method) looks identical to the end of the previous method. Please extract it into a common helper method.

dmitrievanthony replied:
Good idea, done.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h:
```diff
@@ -0,0 +1,77 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
```
This is also missing a header guard.

dmitrievanthony replied:
Added.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h:
```diff
@@ -0,0 +1,77 @@
+  Status RestoreInternal(IteratorContext* ctx,
+                         IteratorStateReader* reader) override;
+
+ private:
```
Style nit: in classes, within each section (`public`, `private` etc.), [put method declarations before data member declarations](https://google.github.io/styleguide/cppguide.html#Declaration_Order).

dmitrievanthony replied:
Done.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc:
```diff
@@ -0,0 +1,147 @@
+
+    const Tensor* permutation_tensor;
+    OP_REQUIRES_OK(ctx, ctx->input("permutation", &permutation_tensor));
+    OP_REQUIRES(ctx, schema_tensor->dims() == 1,
```
s/schema_tensor/permutation_tensor/

dmitrievanthony replied:
Thanks, fixed.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    *output =
+        new IgniteDataset(ctx, cache_name, host, port, local, part, page_size,
```
Use `std::move` for all the non-POD types passed by value here.

dmitrievanthony replied:
Done.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_plain_client_windows.cc:
```diff
@@ -0,0 +1,136 @@
+  for (ptr = result; ptr != NULL; ptr = ptr->ai_next) {
+    sock_ = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol);
+    if (sock_ == INVALID_SOCKET) {
+      WSACleanup();
```
The [documentation](https://docs.microsoft.com/en-us/windows/desktop/api/winsock/nf-winsock-wsacleanup) makes it seem like calling this will terminate other uses of winsock from other threads. Is there a less drastic action?

mrry commented on tensorflow/contrib/ignite/kernels/ignite_plain_client_windows.cc:
```diff
@@ -0,0 +1,136 @@
+    break;
+  }
+
+  freeaddrinfo(result);
```
This is leaked if we return on L62. A `gtl::Cleanup` might be helpful here.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_ssl_wrapper.cc:
```diff
@@ -0,0 +1,142 @@
+
+Status SslWrapper::Disconnect() {
+  SSL_free(ssl_);
+
```
Set `ssl_` to `nullptr` here, to avoid keeping a dangling pointer around. Also consider checking that `ssl_` is nullptr in the destructor.

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+    """Reads the specified number of bytes and returns them as a buffer."""
+    return None
+
+  def __read(self, data_type, length):
```
Style: [Use one leading underscore for private methods.](https://www.python.org/dev/peps/pep-0008/#id47)

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+
+# Binary types defined in Apache Ignite Thin client and supported by
+# TensorFlow on Apache Ignite, see
+# https://apacheignite.readme.io/v2.6/docs/binary-client-protocol.
```
Add a comment to explain what `False`/`True` means here.

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+    """
+    if self.fields is None:
+      object_type = types[self.type_id]
+      if object_type is not None:
```
`object_type` can never be `None`, because `None` is not a value in `types`.

dmitrievanthony replied:
Good point, I added check like:

```python3
if self.type_id in types:
``` 

Done.

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+        if is_array:
+          return tensor_shape.TensorShape([None])
+        return tensor_shape.TensorShape([])
+      raise Exception("Unsupported type [type_id=%d]" % self.type_id)
```
Use `ValueError` instead of `Exception`.

dmitrievanthony replied:
Replaced (where it's appropriate).

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+    return self.to_flat_rec([])
+
+  def to_permutation(self):
+    """Returns a permutation that should be applied to order object leafs."""
```
Typo: s/leafs/leaves/

dmitrievanthony replied:
Thanks, fixed.

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+        name="page_size")
+    self.username = ops.convert_to_tensor("" if username is None else username,\
+        dtype=dtypes.string, name="username")
+    self.password = ops.convert_to_tensor("" if password is None else password,\
```
This (and `self.cert_password`) will encode potentially sensitive information in the `GraphDef` and send it over insecure channels. Is it necessary to include this information here?

dmitrievanthony replied:
It's not necessary, but it's the simple way to start using.

We provide two ways to specify these parameters: via parameters of dataset or via environment variables on the nodes where dataset will be actually instantiated. As far as I understand, in the second case sensitive information won't be included into `GraphDef` and passed via insecure channels.

mrry replied:
Perhaps we should only support the environment variable, in that case? I'm concerned that novice users won't understand the distinction, take the simple route, and leak sensitive information.

However, security isn't my domain, so @martinwicke can you please make a determination here?

mrry commented on tensorflow/contrib/ignite/python/ops/ignite_dataset_ops.py:
```diff
@@ -0,0 +1,763 @@
+    """
+    super(IgniteDataset, self).__init__()
+
+    with IgniteClient(host, port, username, password, certfile, keyfile,\
```
There is no need for a line continuation character here (or in the cases below) because it is implied by the open parentheses.

dmitrievanthony replied:
Thanks, fixed.

mrry reviewed (changes requested):
Note that many of the style/performance/correctness comments apply generally across the code, but I've only commented at the first instance.

dmitrievanthony committed:
Fixes after second review.

dmitrievanthony commented on tensorflow/contrib/ignite/__init__.py:
```diff
@@ -0,0 +1,40 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Apache Ignite is a memory-centric distributed database, caching, and
```
I updated docs and added a short description in the first line.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,272 @@
+limitations under the License.
+==============================================================================*/
+
+#include "ignite_binary_object_parser.h"
```
Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,272 @@
+
+Status BinaryObjectParser::Parse(uint8_t** ptr,
+                                 std::vector<Tensor>* out_tensors,
+                                 std::vector<int32_t>* types) {
```
Yes, excellent point, I completely forgot about it. The reason of `types` to be here is to build a `schema` of every object that we receive and check it. 

I added this logic today, so it affects this code and `CheckTypes` method in `IgniteDatasetIterator`. Please take a look.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.h:
```diff
@@ -0,0 +1,53 @@
+==============================================================================*/
+
+#include <vector>
+#include "tensorflow/core/framework/dataset.h"
```
Updated, look like the following two headers are enough:

```cpp
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/lib/core/status.h"
```

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,272 @@
+      int32_t length = *((int32_t*)*ptr);
+      *ptr += 4;
+      Tensor tensor(cpu_allocator(), DT_STRING, {});
+      tensor.scalar<std::string>()() = std::string((char*)*ptr, length);
```
Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,272 @@
+  switch (object_type_id) {
+    case BYTE: {
+      Tensor tensor(cpu_allocator(), DT_UINT8, {});
+      tensor.scalar<uint8>()() = *((uint8_t*)*ptr);
```
I have to use `reinterpret_cast` in almost all cases because I read data from "generic" byte buffer, but I think I've got the common idea. Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,272 @@
+
+  switch (object_type_id) {
+    case BYTE: {
+      Tensor tensor(cpu_allocator(), DT_UINT8, {});
```
Only one note, compiler requires to use it the following way: `out_tensors->emplace_back(cpu_allocator(), DT_UINT8, TensorShape({}))`. Anyway, done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,272 @@
+    case BYTE_ARR: {
+      int32_t length = *((int32_t*)*ptr);
+      *ptr += 4;
+      Tensor tensor(cpu_allocator(), DT_UINT8, TensorShape({length}));
```
I changed it everywhere on `TensorShape({length})` for consistency. Can't use `{length}`, see the reason above. Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.h:
```diff
@@ -0,0 +1,53 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
```
Header guards added for all headers.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_client.h:
```diff
@@ -0,0 +1,55 @@
+
+  inline Status ReadByte(uint8_t* data) { return ReadData(data, 1); }
+
+  inline Status ReadShort(int16_t* data) { return ReadData((uint8_t*)data, 2); }
```
Yep, good concern. I added today so called `ByteSwapper` that fixes endianness when it's appropriate. Please, take a look.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset.cc:
```diff
@@ -0,0 +1,116 @@
+                             std::vector<int32> schema,
+                             std::vector<int32> permutation)
+    : DatasetBase(DatasetContext(ctx)),
+      cache_name_(cache_name),
```
Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset.cc:
```diff
@@ -0,0 +1,116 @@
+      shapes_.push_back(PartialTensorShape({}));
+    } else if (e >= 12 && e < 21) {
+      shapes_.push_back(PartialTensorShape({-1}));
+    } else {
```
We support hierarchical objects that can be considered as trees. `COMPLEX_OBJ` and `WRAPPED_OBJ` are inner nodes, when all other types are leaves. The `schema`, `output types` and `shapes` are represents only leaves. So, `COMPLEX_OBJ` and `WRAPPED_OBJ` just can't appear here.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset.cc:
```diff
@@ -0,0 +1,116 @@
+    } else if (e == STRING || e == STRING_ARR) {
+      dtypes_.push_back(DT_STRING);
+    } else {
+      LOG(ERROR) << "Unexpected type in schema [type_id=" << e << "]";
```
Good idea, I moved this conversion to `IgniteDataset` and wrapped into `OP_REQUIRES_OK()`. Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+    std::string cert_password, std::vector<int32> schema,
+    std::vector<int32> permutation)
+    : DatasetIterator<IgniteDataset>(params),
+      cache_name_(cache_name),
```
Done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+  if (certfile.empty())
+    client_ = std::unique_ptr<Client>(p_client);
+  else
+    client_ = std::unique_ptr<Client>(new SslWrapper(
```
All things that might fail are placed in `Connect` and `InitSslContext` methods (including access to files), so what you suggest is already implemented actually. `Status` will be returned from `Connect` from the place where this `Connect` is called (`IgniteDatasetIterator::GetNextInternal` generally speaking).

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+Status IgniteDatasetIterator::EstablishConnection() {
+  if (!client_->IsConnected()) {
+    Status status = client_->Connect();
+    if (!status.ok()) return status;
```
Thanks, done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+Status IgniteDatasetIterator::CloseConnection() {
+  if (cursor_id_ != -1 && !last_page_) {
+    Status conn_status = EstablishConnection();
+    if (!conn_status.ok()) return conn_status;
```
Thanks, done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h:
```diff
@@ -0,0 +1,77 @@
+  int32_t JavaHashCode(std::string str) const;
+};
+
+constexpr uint8_t null_val = 101;
```
Looks a bit confusing, but thanks for point. I renamed it, done.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+    int32_t res_len;
+    TF_RETURN_IF_ERROR(client_->ReadInt(&res_len));
+    if (res_len < 12)
+      return errors::Internal("Close Resource Response is corrupted");
```
Done, all `errors::Internal()` replaced with `errors::Unknown()`.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+Status IgniteDatasetIterator::GetNextInternal(IteratorContext* ctx,
+                                              std::vector<Tensor>* out_tensors,
+                                              bool* end_of_sequence) {
+  if (remainder_ == 0 && last_page_) {
```
Done, added mutex-protection.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,437 @@
+    if (remainder_ == -1 || remainder_ == 0) {
+      Status status = remainder_ == -1 ? ScanQuery() : LoadNextPage();
+      if (!status.ok()) return status;
+    }
```
Thanks, really looks clearer. Done.

dmitrievanthony commented:
Hi @mrry. Thank you for very detailed and deep review, I really appreciate it. I think I fixed all your comments today, so could you please have a look my changes?

dmitrievanthony committed:
Add forgotten ignite_byte_swapper.h

dmitrievanthony committed:
Fix windows build.

dmitrievanthony commented:
Meanwhile, @martinwicke, @perfinion, could you please rerun tests? I made a lot of changes, would be great to check CI statuses.

yongtang commented:
: added `kokoro:run` label to help start the tests.

dmitrievanthony committed:
Fix code style.

dmitrievanthony commented:
Thank you, @yongtang. I fixed code style, but I see one confusing thing: `clang-format` suggests another order of "includes". Anyway, I updated the code so that CI checks should pass.

**Regarding Windows builds**, as I wrote above, I see the same issues I had and asked here: https://groups.google.com/a/tensorflow.org/forum/#!topic/build/ePYss0Kxcu4. It looks like we need to add `copt=-DWIN32_LEAN_AND_MEAN` on CI server for Windows builds. So I think the problem on CI-side, not in code. Am I right?

yongtang commented:
@dmitrievanthony I think you could add copts = ["-DWIN32_LEAN_AND_MEAN"] in `tensorflow/contrib/ignite/BUILD` to achieve a similar effect. Maybe you could give it a try?

dmitrievanthony committed:
Add -DWIN32_LEAN_AND_MEAN option into BUILD.

dmitrievanthony commented:
Thanks for a great advice, @yongtang. Looks like it works. Could you please check it on CI?

yongtang commented:
@dmitrievanthony The buildifier to format the BUILD file is:
https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md

yongtang commented:
@dmitrievanthony I created a PR #22258 to fix the Windows GPU build failure issue.

dmitrievanthony committed:
Apply buildifier changes.

dmitrievanthony commented:
Yes, buildifier just wanter reordering, @yongtang. I applied fixes.

dmitrievanthony committed:
Add 'override' specifier to ReadData, WriteData.

dmitrievanthony commented:
Hi @mrry. It's been 5 days since I fixed all review comments. Could you please take a look?

mrry commented on tensorflow/contrib/ignite/kernels/ignite_byte_swapper.h:
```diff
@@ -0,0 +1,129 @@
+ public:
+  ByteSwapper(bool big_endian) {
+    int x = 1;
+    bool is_little_endian = (*(char *)&x == 1);
```
Use the formulation here to work out the endianness statically:

https://github.com/tensorflow/tensorflow/blob/25c99131362f034c3bc3805d741f0c4ab9d0cb8b/tensorflow/compiler/xla/literal.cc#L46

dmitrievanthony replied:
I updated it similar way.

dmitrievanthony committed:
Work out the endianness statically.

dmitrievanthony commented:
Regarding security question, @mrry, @martinwicke, I think we can add warning in case user specifies sensitive data via parameters. What do you think?

martinwicke commented:
I think we should make it hard for credentials to end up in files or on the network in the clear. Therefore, I would prefer if we insist here that all credentials are present on the executing machine already, and we make it impossible to use credentials that are stored in the graph. 

I think environment variables would work fine in this case, can we restrict it to that for all sensitive information?

dmitrievanthony committed:
Avoid saving sensitive information in graph.

dmitrievanthony commented:
I agree, it's reasonable, @mrry, @martinwicke. I updated code so that sensitive information is not encoded in graph. Be aware that it's still used in python for initial access (that is required to get data schema), but after that only environment variables are used.

dmitrievanthony commented:
Do we have any open questions, @mrry?

mrry reviewed (dismissed):
Let's go ahead and merge this.

dmitrievanthony committed:
Fix clang styles.

dmitrievanthony committed:
Fix byte-order issue.

dmitrievanthony commented:
Guys, looks like I introduced a bug on Windows when changed `ignite_byte_swapper.h`. It's fixed now, so please rerun tests who can do it.

yongtang commented:
@dmitrievanthony The windows GPU fix has been submitted internally so the build should pass. Let me help with rerun the test.

dmitrievanthony commented:
Ok, @mrry, @martinwicke, @yongtang, all tests I expected to pass actually passed. XLA fails, but it fails all the time by reasons that are not related to my code.

Derek, sorry for bothering you again, but we need you approval again. After that this PR can be merged, right?

mrry reviewed (approved):
Re-approving. There are a few more internal steps required before the PR is merged, but you shouldn't need to do anything else from this point.

dmitrievanthony commented:
Hi, @mrry, @martinwicke. I'm not sure how long it takes to pass all internal steps, but it looks like it takes time. Meanwhile, as far as I see, there are several conflicts appeared. Shall I fix them?
